File: Friends Pairing Problem.java
class Solution
{
    public long countFriendsPairings(int n) 
    { 
       //code here
       if(n<=2)return n;
       long mod=(int)Math.pow(10,9)+7;
       long secPrev=1;
       long prev=2;
       long curr=0;
       for(int i=3;i<=n;i++){
           curr=(prev+(((i-1)%mod)*(secPrev%mod))%mod)%mod;
           secPrev=prev;
           prev=curr;
       }
       return curr;
    }
}    

File: Friends Pairing Problem.cpp
class Solution
{
public:

    //defining mod
    long long MOD = 1000000007;
    
    int countFriendsPairings(int n) 
    { 
        //numbers less than 3 will have the same answer
        if(n < 3) return n;
        
        //defining and storing base conditions
        long long prev1 = 1;
        long long prev2 = 2;
        
        //using tabulation based on the pattern observed
        for(long long i = 3; i <= n; i++)
        {
            long long temp = prev2 % MOD;
            
            prev2 = (((((i-1) % MOD)*(prev1 % MOD))%MOD) + (prev2 % MOD))%MOD;
            prev1 = temp;
        }
        
        return prev2;
    }
};
